### PR TITLE
[RATOM-69] DRF-ification

### PR DESCRIPTION
**Pull request by michael-caktus**
The server side work of moving from GraphQL to REST.

Refactors existing state of API using DRF instead of Graphene.

Some of that work includes:
Sets up MessageDocument and MessageView, with django-elasticasearch-dsl-drf
Adds serializers
Sets up DRF jwt auth

### CODE REVIEW

File: ratom_api/settings/base.py [hunk 1]
```diff
@@ -26,6 +15,7 @@
 
 INSTALLED_APPS = [
     "django.contrib.auth",
+    "corsheaders",
```
Comment (copelco) [inline ratom_api/settings/base.py]: New requirement needed in requirements/base.txt, I think

File: ratom_api/settings/base.py [hunk 2]
```diff
@@ -34,14 +24,18 @@
     "django.contrib.humanize",
     "django.contrib.sitemaps",
     "django.contrib.postgres",
+    "rest_framework",
```
Comment (copelco) [inline ratom_api/settings/base.py:26]: Same here

File: ratom_api/settings/base.py [hunk 3]
```diff
@@ -34,14 +24,18 @@
...
+    # Django Elasticsearch integration
     "django_elasticsearch_dsl",
-    "graphene_django",
+    # Django REST framework Elasticsearch integration (this package)
+    "django_elasticsearch_dsl_drf",
```
Comment (copelco) [inline ratom_api/settings/base.py:28]: Same here

File: ratom_api/settings/base.py [hunk 4]
```diff
@@ -50,28 +44,51 @@
...
+    "origin",
+    "user-agent",
+    "x-csrftoken",
+    "x-requested-with",
+    "x-version-requested",
+]
+
 AUTHENTICATION_BACKENDS = [
-    "graphql_jwt.backends.JSONWebTokenBackend",
+    # "graphql_jwt.backends.JSONWebTokenBackend",
```
Comment (copelco) [inline ratom_api/settings/base.py]: I think we should probably remove any references to graphql

File: ratom_api/urls.py [hunk 5]
```diff
@@ -1,35 +1,11 @@
...
 
 urlpatterns = [
-    path('admin/', admin.site.urls),
-    path('graphql', csrf_exempt(GraphQLView.as_view(graphiql=True, schema=schema))),
-] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
-
-if settings.DEBUG:
-    import debug_toolbar
-    urlpatterns += [
-        re_path(r'^__debug__/', include(debug_toolbar.urls)),
```
Comment (copelco) [inline ratom_api/urls.py:34]: debug-toolbar may still be useful debugging query performance on the REST views. Do we want to remove it?

File: ratom_api/_schema.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# import graphene
```
Comment (copelco) [inline ratom_api/_schema.py]: Delete

File: ratom/_documents.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
+# from typing import List
```
Comment (copelco) [inline ratom/_documents.py]: Delete

File: ratom/_schema.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
+# from datetime import datetime
```
Comment (copelco) [inline ratom/_schema.py]: Delete

File: ratom/importer.py [hunk 9]
```diff
@@ -88,44 +88,42 @@ def _create_message(
             return None
         # spaCy jsonb (idea #1)
         labels = set()
-        for entity in document.ents:
-            labels.add(entity.label_)
+        # for entity in document.ents:
```
Comment (copelco) [inline ratom/importer.py]: I believe we still need these two lines, otherwise the import won't load any tags for elasticsearch to use.

**Review by copelco — CHANGES_REQUESTED** [review]
Added some initial comments. Excited to test it out.

File: ratom_api/settings/base.py [hunk 10]
```diff
@@ -50,28 +40,50 @@
     "django.middleware.clickjacking.XFrameOptionsMiddleware",
 ]
 
+CORS_ORIGIN_ALLOW_ALL = True
```
Comment (copelco) [inline ratom_api/settings/base.py]: We should probably move these 2 settings to ``dev.py`` too

File: ratom/admin.py [hunk 11]
```diff
@@ -1,37 +1,24 @@
...
-@admin.register(Entity)
-class EntityAdmin(admin.ModelAdmin):
-    list_display = (
-        "pk",
-        "label",
-        "value",
-    )
-    list_filter = ("label",)
-    search_fields = ("value",)
+    # list_filter = ("sent_date", "collection")
```
Comment (copelco) [inline ratom/admin.py]: Probably should re-enable most of these since they're helpful in the admin

File: ratom/permissions.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+from rest_framework import permissions
+from .models import User
+
+
+class IsOwnerOrReadOnly(permissions.BasePermission):
```
Comment (copelco) [inline ratom/permissions.py]: Is this class used?

File: ratom/models.py [hunk 13]
```diff
@@ -32,46 +33,35 @@ class Processor(models.Model):
...
+        Example:
+        >>> mapping = {
+        >>>     "account": {
+        >>>         "title": "Gov Purdue"
+        >>>     }
+        >>> }
+
+        :return:
+        """
+        return dict_to_obj({"title": self.account.title,})
```
Comment (copelco) [inline ratom/models.py:59]: 🆒

File: ratom/models.py [hunk 14]
```diff
@@ -32,46 +33,35 @@ class Processor(models.Model):
...
+
+    @property
+    def labels_indexing(self):
+        labels = []
+        if self.data:
+            labels = list(self.data.get("labels", []))
+        return labels
+        # return [tag.title for tag in self.tags.all()]
```
Comment (copelco) [inline ratom/models.py]:
```suggestion

```

File: ratom/views.py [hunk 15]
```diff
@@ -1,3 +1,178 @@
...
+        # TODO: need to decide how/if to limit access
+        accounts = Account.objects.all()
+        serialized_account = AccountSerializer(accounts, many=True)
+        return Response(serialized_account.data)
+
+    if request.method == "POST":
+        serialized_account = AccountSerializer(data=request.data)
+        if serialized_account.is_valid():
+            serialized_account.save()
+            return Response(serialized_account.data, status=status.HTTP_201_CREATED)
```
Comment (copelco) [inline ratom/views.py:54]: Now is probably a good time to agree on what we should be unit testing. Ideally all of these views would include unit tests. Happy to help write some up.

File: ratom/views.py [hunk 16]
```diff
@@ -1,3 +1,178 @@
...
+        return Response(status=status.HTTP_404_NOT_FOUND)
+
+    if request.method == "GET":
+        serialized_user = UserSerializer(user)
+        return Response(serialized_user.data)
+
+
+@api_view(["GET", "POST"])
+@permission_classes([IsAuthenticated])
+def account_list(request):
```
Comment (copelco) [inline ratom/views.py:39]: I don't think we need to change, but I wonder what the class-based views would look like here: https://www.django-rest-framework.org/api-guide/generic-views/#concrete-view-classes

File: ratom/views.py [hunk 17]
```diff
@@ -1,3 +1,178 @@
...
+        message = Message.objects.get(pk=pk)
+    except Message.DoesNotExist:
+        return Response(status=status.HTTP_404_NOT_FOUND)
+
+    if request.method == "GET":
+        serialized_message = MessageSerializer(message)
+        return Response(serialized_message.data)
+
+
+HIGHLIGHT_TAGS = {
```
Comment (copelco) [inline ratom/views.py:100]: I don't see highlights in my search currently - should I?

**Review by copelco — COMMENTED** [review]
Looks cool. Pushed up a few small changes. Added a few more comments.

File: ratom/views.py (hunk 16 above)
Comment (michael-caktus) [reply ratom/views.py:39]: I'm totally open to making the change, if we want to. Using function based views feels more easily customizable to me, but that's probably just a comfort-level thing.

**Review by copelco — APPROVED** [review]

**Pull request merged**